File: RmaxminOnJava/Rmaxmin.java
package ru.prinkov.lab;

public class Rmaxmin
{
	private double value;	
	private static final double ZERO = Double.NEGATIVE_INFINITY;
	private static final double UNIT = Double.POSITIVE_INFINITY; 

	public Rmaxmin()
	{
		this.value = Rmaxmin.ZERO;
	}

	public Rmaxmin(double value_)
	{
		this.value = value_;
	}

	public Rmaxmin(Rmaxmin other)
	{
		this.value = other.value;
	}

	public Rmaxmin add(Rmaxmin other)
	{
		return new Rmaxmin(max(this.value, other.value));
	}

	public Rmaxmin multiply(Rmaxmin other)
	{
		return new Rmaxmin(min(this.value, other.value));
	}

	public String toString()
	{
		if(value == Double.POSITIVE_INFINITY)
			{
				return "+oo";
			}
		else if(value == Double.NEGATIVE_INFINITY)
			{
				return "-oo";
			}
		return ""+value;
	}

	private double min(double a, double b)
	{
		if(a < b) return a;
			else return b;
	}

	private double max(double a, double b)
	{
		if(a > b) return a;
			else return b;
	}
}

File: Rmax/mnozhestvo/Rmaxmin.cpp
#include "Rmaxmin.hpp"
	Rmaxmin::Rmaxmin()
	{
		value = -std::numeric_limits<double>::infinity();
	}
	Rmaxmin::Rmaxmin(double param)
	{
		value = param;
	}

	Rmaxmin::Rmaxmin(Rmaxmin &param)
	{
		value = param.value;
	}

	Rmaxmin Rmaxmin::zero()
	{
		Rmaxmin temp(-std::numeric_limits<double>::infinity());
		return temp;
	}

	Rmaxmin Rmaxmin::unit()
	{
		Rmaxmin temp(std::numeric_limits<double>::infinity());
		return temp;
	}

	std::string Rmaxmin::toString()
	{
		std::stringstream ostr;
		if (value == std::numeric_limits<double>::infinity())
			ostr << "oo";
		else if (value == -std::numeric_limits<double>::infinity())
			ostr << "-oo";
		else
			ostr << value;
		return ostr.str();
	}

	std::istream & operator >> (std::istream & s, Rmaxmin &n)
	{
			s >> n.value;
			return s;
	}

	std::ostream & operator <<(std::ostream & s, const Rmaxmin &n)
	{
		s << n.toString();
		return s;
	}

	Rmaxmin Rmaxmin::operator+(const Rmaxmin& other)
	{
		Rmaxmin temp;
		value>b.value?temp.value=value:temp.value=b.value;
		return temp;
	}

	Rmaxmin Rmaxmin::operator*(const Rmaxmin& other)
	{
		Rmaxmin temp;
		value<b.value ? temp.value = value : temp.value = other.value;
		return temp;
	}

	Rmaxmin Rmaxmin::operator^(const Rmaxmin& other)
	{
		Rmaxmin temp(*this);
		return temp;
	}
	
	Rmaxmin& Rmaxmin::operator=(const Rmaxmin &b)
	{
		
		if (*this==b) return *this;
		this -> value = other.value;
		return * this;
		
	}

	Rmaxmin& Rmaxmin::operator=(const double &b)
	{

		if (*this == other) return *this;
		this->value = other;
		return *this;

	}

	bool operator ==(const Rmaxmin & a, const Rmaxmin & other)
	{
		if (!(a.value == other.value))     return false;
		return true;
	}
